spatial/transform/umeyama.go

```diff
@@ -0,0 +1,228 @@
+//
+// Reference:
+// "Least-Squares Estimation of Transformation Parameters Between Two Point Patterns"
+// by Shinji Umeyama, IEEE Transactions on Pattern Analysis and Machine Intelligence,
+// Vol. 13, No. 4, April 1991, [doi:10.1109/34.88573].
+// [doi:10.1109/34.88573]: https://doi.org/10.1109/34.88573
+type Umeyama struct {
+	x, y     *mat.Dense
+	n, m     int
+	muX, muY *mat.VecDense
+	varX     float64
+}
+
+// NewUmeyama creates a new Umeyama similarity transformation calculator
+// for the given point sets x and y.
+//
+// The point sets are represented as two n×m matrices X and Y, where
+// m is the number of dimensions and each row represents a point.
+// If the dimensions of X and Y are not equal, NewUmeyama will panic.
+//
+// NewUmeyama computes the means and variance of the input points but does
+// not perform the full transformation calculation. Use the Var method to
+// inspect the variance, and the Transform method to compute the transformation
+// parameters.
+func NewUmeyama(x, y *mat.Dense) *Umeyama {
+	n, m := x.Dims()
+	rowsY, colsY := y.Dims()
+
+	// Check dimensions.
+	if n != rowsY || m != colsY {
+		panic("transform: dimensions of x and y do not match")
+	}
+
+	// Calculate means and variance of x.
+	muX := mat.NewVecDense(m, nil)
+	muY := mat.NewVecDense(m, nil)
+
+	colX := make([]float64, n)
+	colY := make([]float64, n)
+
+	var varX float64
+
+	for j := 0; j < m; j++ {
+		mat.Col(colX, j, x)
+		mat.Col(colY, j, y)
+
+		meanX, varXj := stat.PopMeanVariance(colX, nil)
+
+		muY.SetVec(j, stat.Mean(colY, nil))
+		muX.SetVec(j, meanX)
+
+		varX += varXj
+	}
+
+	return &Umeyama{
+		x:    x,
+		y:    y,
+		n:    n,
+		m:    m,
+		muX:  muX,
+		muY:  muY,
+		varX: varX,
+	}
+}
+
+// Var returns the variance of point set x.
+//
+// This can be used to detect degenerate input cases where the variance
+// is too low, which may cause numerical instability and/or division by zero
+// in the transformation calculation.
+func (u *Umeyama) Var() float64 {
+	return u.varX
+}
+
+// Transform computes and returns the similarity transformation parameters.
+//
+// Transform returns the scale factor c, the rotation matrix R and the translation
+// vector t that best align the point sets according to Umeyama's algorithm.
+//
+// If the required SVD fails, Transform will return a mat.ErrFailedSVD.
+func (u *Umeyama) Transform() (c float64, r *mat.Dense, t *mat.VecDense, err error) {
+	// Center the matrices.
+	xc := mat.NewDense(u.n, u.m, nil)
+	yc := mat.NewDense(u.n, u.m, nil)
+
+	for i := 0; i < u.n; i++ {
+		for j := 0; j < u.m; j++ {
+			xc.Set(i, j, u.x.At(i, j)-u.muX.AtVec(j))
+			yc.Set(i, j, u.y.At(i, j)-u.muY.AtVec(j))
+		}
+	}
+
+	// Calculate covariance matrix.
+	covXY := mat.NewDense(u.m, u.m, nil)
+	covXY.Mul(yc.T(), xc)
+	covXY.Scale(1/float64(u.n), covXY)
+
+	// Singular Value Decomposition
+	var svd mat.SVD
+	if !svd.Factorize(covXY, mat.SVDFull) {
+		return 0, nil, nil, mat.ErrFailedSVD
```

Comment:
With the new design, the only failure that can happen in this method is the SVD factorisation. So we don't need an error at all; this can just return an `ok bool` like the actual factorisation does.